src/Cooked/MockChain/Balancing.hs

```diff
@@ -1,409 +1,326 @@
-      -- present on the transaction.
-      inputOrefs = Map.keys $ txSkelIns skel
-  -- Get all UTxOs that belong to the given wallet, and that are not yet being
-  -- consumed on the transaction.
-  --
-  -- These UTxOs are sorted in decreasing order of their Ada value, which will
-  -- make 'selectNewInputs' will more likely select additional inputs that
-  -- contain a lot of Ada. The hope behind this heuristic is that it'll
-  -- therefore become less likely for the 'returnValue' to be less than the
-  -- minimum Ada amount required for each output. See this comment for context:
-  -- https://github.com/tweag/cooked-validators/issues/71#issuecomment-1016406041
-  candidateUtxos <-
-    sortBy (flip compare `on` Script.fromValue . outputValue . snd)
-      . filter ((`notElem` inputOrefs) . fst)
-      <$> utxosAt (walletAddress balanceWallet)
-  case selectNewInputs candidateUtxos [] initialExcess missingValue of
-    Nothing ->
-      throwError $
-        MCEUnbalanceable
-          (MCEUnbalNotEnoughFunds balanceWallet missingValue)
-          skel
-    Just bTxRes -> return bTxRes
-  where
-    selectNewInputs ::
-      [(Api.TxOutRef, Api.TxOut)] ->
-      [(Api.TxOutRef, Api.TxOut)] ->
-      Api.Value ->
-      Api.Value ->
-      Maybe BalanceTxRes
-    selectNewInputs available chosen excess missing =
-      case view flattenValueI missing of
-        [] -> Just $ BalanceTxRes chosen excess available
-        (ac, _) : _ ->
-          -- Find the first UTxO belonging to the wallet that contains at least
-          -- one token of the required asset class (The hope is that it'll
-          -- contain at least @n@ such tokens, but we can't yet fail if there
-          -- are fewer; we might need to add several UTxOs):
-          case break ((`Script.geq` Script.assetClassValue ac 1) . outputValue . snd) available of
-            (_, []) -> Nothing -- The wallet owns nothing of the required asset class. We can't balance with this wallet.
-            (previousUtxos, theChosenUtxo : nextUtxos) ->
-              let available' = previousUtxos ++ nextUtxos
-                  chosen' = theChosenUtxo : chosen
-                  theChosenValue = outputValue $ snd theChosenUtxo
-                  theChosenDifference = missing <> PlutusTx.negate theChosenValue
-                  excess' = excess <> negativePart theChosenDifference
-                  missing' = positivePart theChosenDifference
-               in -- A remark on why the following line should not lead to an
-                  -- infinite recursion: The value described by @missing'@ is
-                  -- strictly smaller than the value described by @missing@,
-                  -- because there was at least one token of the asset class
-                  -- @ac@ in @theChosenValue@.
-                  selectNewInputs available' chosen' excess' missing'
+-- | This reduces a set of given collateral inputs while accounting for:
```

Comment:
I am not sure I understand the intended meaning of "reduces" (here and the rest of the module). This means minimizing the cardinal of the set or the total amount of value they contain?